Vamsi krishna creates a stacked horizontal bar graph for tools stoppage reason

vamsikrishna1704 opened the pull request:
# Description
<img width="770" height="765" alt="image" src="https://github.com/user-attachments/assets/50103158-ebdb-46b3-b4e4-e68144d076e5" />


## Related PRS (if any):
This frontend PR is related to the [1550](https://github.com/OneCommunityGlobal/HGNRest/pull/1550) backend PR.
…

## Main changes explained:
- Created ToolsStoppageHorizontalBarChart.jsx file to implement the stacked horizontal bar chart of tools stoppage reason.
- Update WeeklyProjectSummary.jsx to include ToolsStoppageHorizontalBarChart component.
- Create ToolsStoppageHorizontalBarChart.css file to provide styling to the stacked horizontal bar chart.
…

## How to test:
1. check into current branch
2. do `npm install` and `npm run start:local` to run this PR locally
3. Clear site data/cache
4. log as admin user
5. go to dashboard→ Reports→ Total Construction Summary→ Tools and Equipment Tracking Card
6. verify the filters to see appropriate results.

## Screenshots or videos of changes:
Walkthrough - light theme
https://github.com/user-attachments/assets/789e07ec-8dcf-4a61-bb6b-541476e99c37

Dark Theme:
https://github.com/user-attachments/assets/a2165a68-61a9-4f42-a587-6251939338c8

vamsikrishna1704 committed:
Created the Tools Stoppage Horizontal Bar Chart page.

vamsikrishna1704 committed:
Fixed issues rose on connecting frontend to backend api

vamsikrishna1704 committed:
Fixed the BG of recharts tooltip cursor of darkmode

vamsikrishna1704 committed:
Updated the styles for tooltip content

beblicarl commented:
This feature works as intended and the code is great

https://www.loom.com/share/f7064847b0e6449490b6c5c527584057

khan-zoha reviewed (changes requested):
Reviewed PR 3780. The node version is not compatible with node 20.
<img width="810" height="214" alt="Screenshot 2025-09-13 at 7 22 51 PM" src="https://github.com/user-attachments/assets/cd492f7f-9385-440c-ae7e-90b2d581f8a2" />

vamsikrishna1704 committed:
Node version upgrade issues fixed.

vamsikrishna1704 committed:
Node version upgrade issues fixed.

laynet reviewed (changes requested):
Getting an error that this branch is not compatible with node v 20
![PR 3780](https://github.com/user-attachments/assets/438f4f55-027d-45e7-a389-188b6eba89ed)

MeronTeweldebrhan reviewed (changes requested):
PR Review Feedback

Testing Outcome

✅ Verified Tools and Equipment Tracking Card under Dashboard → Reports → Total Construction Summary.
✅ Proportion of Tools/Equipment filter works as expected.
✅ Tools by Availability chart functions correctly.
✅ Dark mode verified.
❌ Reason of Stoppage of Tools fails to load, showing: “Failed to load projects. Please try again.”

Suggestion
In the Proportion of Tools/Equipment filter, after selecting a project, switching directly to another project doesn’t refresh the list unless “All” is clicked first. Improving this behavior would enhance user experience.

<img width="1893" height="834" alt="Screenshot 2025-09-25 112806" src="https://github.com/user-attachments/assets/a4631398-31ab-4511-8f3b-04e58fb3f926" />
<img width="1915" height="833" alt="Screenshot 2025-09-25 112823" src="https://github.com/user-attachments/assets/05a40067-610e-4dbe-8b03-b26b469d3d70" />

vamsikrishna1704 committed:
Resolved the merge conflicts with latest yarn.lock

akshith312 reviewed (dismissed):
Tested Locally. 
Logged in as Admin User and navigated to page.
Filters and Buttons work as expected in both Light and Dark Modes.
<img width="1916" height="945" alt="image" src="https://github.com/user-attachments/assets/06b68d2a-5b87-4173-8e9f-80e270078bed" />
<img width="1918" height="952" alt="image" src="https://github.com/user-attachments/assets/37fb0326-f8ec-4c62-b1d5-62990a77df4d" />
<img width="1917" height="957" alt="image" src="https://github.com/user-attachments/assets/eda6fb68-a81b-4a59-af46-70e228d85be2" />

sumedhkumar96 reviewed (approved):
**Followed the instructions provided by the dev. The summary dashboard, along with its filters and visualizations, works as intended. All the components are clearly visible in the dark mode as well.** 

<img width="1920" height="977" alt="Screenshot 2025-10-09 195105" src="https://github.com/user-attachments/assets/61b7b421-9def-4619-ab9b-9f2c7c9f39b0" />
<img width="1920" height="973" alt="Screenshot 2025-10-09 203333" src="https://github.com/user-attachments/assets/036bf4bb-2062-40d5-b5c4-463dcd9021bf" />

ManeeshBuddha21 commented on src/components/BMDashboard/Tools/ToolsStoppageHorizontalBarChart.jsx:
```diff
@@ -0,0 +1,279 @@
+  const [error, setError] = useState(null);
+  const [data, setData] = useState([]);
+  const emptyData = [];
+
```
just a minor suggestion, its better if we Define emptyData outside the component to avoid re-creation on every render.

ManeeshBuddha21 commented on src/components/BMDashboard/Tools/ToolsStoppageHorizontalBarChart.module.css:
```diff
@@ -0,0 +1,118 @@
+  border-radius: 4px;
+}
+
+:global(.darkTheme .form-control .datePickerInput) {
```
Styling looks great! I saw some :global rules ,  are those needed here or could local classes work too? Just curious about the convention.

ManeeshBuddha21 commented on src/components/BMDashboard/Tools/ToolsStoppageHorizontalBarChart.jsx:
```diff
@@ -0,0 +1,279 @@
+      backgroundColor: state.isFocused ? '#364156' : '#2c3344',
+      color: '#e0e0e0',
+    }),
+    singleValue: base => ({
```
Styling looks great! I noticed some hard-coded colors here , are these intended to match the dark theme or could we pull them from global theme variables later? Just asking to understand the convention.

ManeeshBuddha21 reviewed (commented):
Great work on this feature ,the code is really well structured, and the chart logic is clear and easy to follow.
I’ve left a few minor notes around color consistency, effect cleanup, and style conventions , mostly small suggestions and learning observations.

I wasn’t able to verify the actual UI locally due to some setup issues, but I reviewed the code carefully based on what I understood. Everything looks clean overall, and my comments are just small improvement ideas.

Vinay944924 commented:
Hi Vamsi,
I tested the PR lcoally, it seems to work fine in both dark and light mode but it showed an error of failed to load the projects. Can you please verify this
<img width="1468" height="870" alt="Screenshot 2025-11-04 at 7 30 44 PM" src="https://github.com/user-attachments/assets/510efab2-ad2b-4a81-a7e3-ac3c2bf0d1a0" />
<img width="1468" height="870" alt="Screenshot 2025-11-04 at 7 30 53 PM" src="https://github.com/user-attachments/assets/af280379-500a-4b0c-94e9-5fd34c100947" />

Vinay944924 reviewed (changes requested):
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/3780#issuecomment-3488561423

shashankm9 committed:
Create stacked horizontal bar graph for tools stoppage reason

- Add backend API controller for tools stoppage data aggregation
- Implement MongoDB aggregation pipeline for percentage calculations
- Create React component with Chart.js for horizontal stacked bars
- Add project and date range filtering functionality
- Include proper color coding and data labels
- Support dark mode and responsive design

shashankm9 committed:
Refactor ToolsStoppageHorizontalBarChart to reduce cognitive complexity

- Extract processResponseData helper function
- Extract fetchStoppageDataForProject helper function
- Simplify main useEffect logic to reduce nesting
- Reduce cognitive complexity from 16 to under 15 (SonarQube requirement)

shashankm9 committed:
Comprehensive refactor to reduce cognitive complexity below 15

- Extract complex conditions into helper functions (shouldShowChart, shouldShowEmptyState)
- Break down handleDataFetching into smaller single-responsibility functions
- Add early returns to avoid deep nesting
- Extract chart configuration logic (getChartData, getChartOptions)
- Split date formatting and response handling into separate functions
- Apply all SonarQube cognitive complexity best practices

shashankm9 committed:
Fix: Reduce handleDataFetching parameters from 8 to 1 object parameter

- Group related parameters into a single object parameter
- Fixes SonarQube rule: max 7 parameters for functions
- Use destructuring assignment for clean parameter access

shashankm9 committed:
Fix light mode calendar styling and use httpService for auth

shashankm9 committed:
fix: dark mode alignment for calendar filter and project dropdown

shashankm9 committed:
Fix: center-align and style No data available text in dark mode

shashankm9 committed:
retrigger sonar

one-community committed:
Merge pull request #4376 from OneCommunityGlobal/fix_tools_stoppage_reason

Mani shashank fix stacked horizontal bar graph for tools stoppage reason
